### ISSUE TITLE
Iceberg sink for Apache Flink

### ISSUE THREAD

**Issue by bowenli86** [apache/iceberg#567]
Users should be able to have their Flink apps write into Iceberg directly via a sink to leverage Iceberg  as the table format for their near-real-time or even real time data lake/warehouse.

Background: Apache Flink is the most popular and powerful stream processing engine nowadays, and has been widely adopted in companies like Uber, Lyft, Netflix, etc.

**Comment by bowenli86** [comment]
Hi @rdblue , I wonder if this is something that the iceberg community will consider?

**Comment by rdblue** [comment]
Yeah, we (Netflix) have a prototype sink from Flink to Iceberg that we use. I think it's a great idea to add an open one; we've been meaning to clean up ours if we can to get it upstream. Would you want to add a sink in Flink or in Iceberg?

**Comment by bowenli86** [comment]
I'd like to do it in Iceberg first. When it and the Iceberg project matures, I can help evaluate merging it back to Flink.

Would you prefer to use your internal one, or develop a new one from scratch (which I can help with)? Either way, it would be great if you can share some code snippets of the prototype sink early to gather some feedback and ideas.

**Comment by rdblue** [comment]
I don't think that it makes sense to keep code in one repository or another based on incubation status. I'm wondering how difficult it would be to maintain the code here vs in Flink. Does Flink have a stable connector API for plugging in sinks?

**Comment by bowenli86** [comment]
Sorry for the confusion. I was not suggesting incubating/graduating as an argument, but only an arbitrary time in the future to reevaluate the situation.

I think we better do it in iceberg at least for now because 1) lt potentially can take longer time (google pubsub connector took ~a year, pulsar connector took ~1.5 y and hasn't get in yet) in Flink to reach consensus, write FLIP (Flink improvement proposal) and get it reviewed, etc, and I don't know if someone wants to drive that 2) AFAIU, Flink tends to have less connectors in its repo (but show them on Flink website) to reduce build/test time 3) iceberg itself can potentially evolve faster as an emerging tech and it can be easier to make frequent changes in itself

In a nutshell, I feel it's much easier and more efficient if we do it in iceberg and help the community to grow.

**Comment by jerryshao** [comment]
Just curious about the status of Flink connector. Iceberg requires lots of things exposed from connector, so that Iceberg could leverage it.

**Comment by stevenzwu** [comment]
As for open sourcing Flink Iceberg sink that @rdblue mentioned, it seems that we probably can only get to it on later Q1 or early Q2 due to resource constrains. Need to figure out how to untangle the Netflix specific logics. Goal is to make the open-source Flink sink connector generic and extensible

**Comment by aokolnychyi** [comment]
[Here](https://lists.apache.org/thread.html/73585fdfc740a63041eb60ccd71c80b9d6c34c97cfe16a871bc93983@%3Cdev.iceberg.apache.org%3E) is a relevant thread on the dev list.

**Comment by bowenli86** [comment]
@stevenzwu look forward to that. Thanks!

**Comment by waterlx** [comment]
Working on porting/moving the code provided as [nfflink-connector-iceberg](https://github.com/Netflix-Skunkworks/nfflink-connector-iceberg) to master. Hopefully could make it done and upload a PR in the middle of next week. It might be useful if anyone would like to get a simple working version of Flink Iceberg sink for PoC. @stevenzwu @bowenli86  FYI

**Comment by bowenli86** [comment]
@waterlx awesome, thanks! Let us know if you need any help.

**Comment by openinx** [comment]
Thanks @waterlx for your work, would take a look if the PR is prepared.

**Comment by waterlx** [comment]
Just would like to share some progress:

I have moved the source code part of [nfflink-connector-iceberg](https://github.com/Netflix-Skunkworks/nfflink-connector-iceberg) into a dedicate project (iceberg-flink) under Iceberg, also made some corrections and improvements, as well as made most of the checkstyle issues cleaned.   

It works (DataFiles generated and commit made as expected) but is still an immature version. 
We also found some bugs (such as duplicate records) as well as some enhancments good to have(such as if we could remove the usage of Avro IndexedRecord as the intermediate format). 

Plan to upload a WIP PR in the next few days (hopefully before the end of this week) to show the current workable version. 
FYI @stevenzwu @bowenli86 @openinx @aokolnychyi @chenjunjiedada @rdblue

**Comment by openinx** [comment]
@waterlx , I think the first pull request version don't have to be so perfect, can put the rough patch to review and see what's the feedback from community, and create the following issues to address them :-)

**Comment by waterlx** [comment]
Here comes the WIP PR for Flink Iceberg sink: https://github.com/apache/incubator-iceberg/pull/856.
@bowenli86 @openinx @aokolnychyi @chenjunjiedada @jerryshao @stevenzwu 
It is a workable versio, and needs to be improved and polished.
I will sum up a users' guide and the to-do list later.

**Comment by waterlx** [comment]
The design doc extracted as well as future work (still being collected) and sample code is attached [here](https://docs.google.com/document/d/19M-sP6FlTVm7BV7MM4Om1n_MVo1xCy7GyDl_9ZAjVNQ/edit?usp=sharing), in case you are interested.

**Comment by waterlx** [comment]
Updated the design doc authority to allow comments, thanks for the reminder Ryan!

**Comment by stevenzwu** [comment]
@waterlx sorry for being late. didn't get any notification email from github. I will try to comment in the individual jiras to provide more context.

**Comment by openinx** [comment]
Since we've merged the flink sink connector, it's time to close this issue now.
